OptimizelySDK/Notification/NotificationBroadcaster.cs

```diff
@@ -0,0 +1,120 @@
+﻿/* 
+ * Copyright 2017, Optimizely
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using OptimizelySDK.Entity;
+using OptimizelySDK.Event;
+using OptimizelySDK.Logger;
+using System.Collections.Generic;
+
+namespace OptimizelySDK.Notification
+{
+    /// <summary>
+    /// Manages Optimizely SDK notification listeners and broadcasts messages to listeners.
+    /// </summary>
+    public class NotificationBroadcaster
+    {
+        private ILogger Logger = new DefaultLogger();
+
+        private List<NotificationListener> NotificationListeners =
+            new List<NotificationListener>();
+
+        /// <summary>
+        /// Add notification listener if it does not exist already.
+        /// </summary>
+        /// <param name="notificationListener">The notification listener</param>
+        public void AddListener(NotificationListener notificationListener)
+        {
+            if (notificationListener == null)
+            {
+                Logger.Log(LogLevel.ERROR, "Notification listener must not be null.");
+                return;
+            }
+
+            if (NotificationListeners.Contains(notificationListener))
+            {
+                Logger.Log(LogLevel.DEBUG, "Notification listener cannot be added because it already exists.");
+                return;
+            }
+
+            NotificationListeners.Add(notificationListener);
+            Logger.Log(LogLevel.INFO, "Notification listener has been added.");
+        }
+
+        /// <summary>
+        /// Remove notification listener if exists.
+        /// </summary>
+        /// <param name="notificationListener">The notification listener</param>
+        public void RemoveListener(NotificationListener notificationListener)
+        {
+            if (notificationListener == null)
+            {
+                Logger.Log(LogLevel.ERROR, "Notification listener must not be null.");
+                return;
+            }
+
+            if (!NotificationListeners.Contains(notificationListener))
```

Comment:
how .Contains can check listener. with reference type. looks weird to me.

Reply:
Lol ... Looks equally weird to me when I was implementing the class. But in C#, specially in the case of collections, it's almost a universally accepted rule that if we do not implement the equality comparison methods, many things will not gonna work as expected. 

It is the responsibility of class creator to handle such scenarios. That's why I leave the interface as it is. We can restrict equality comparison by adding IEquals interface in the abstract claaa I think.

Btw, in case of default reference equality comparison neither .Contains nor .Remove methods can work as expected.